hs-abci-sdk/src/Tendermint/SDK/Modules/Validators/Keeper.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Tendermint.SDK.Modules.Validators.Keeper where
+
+import qualified Data.Map.Strict                         as Map
+import           Data.Maybe                              (fromMaybe)
+import qualified Data.Set                                as Set
+import           Data.Word                               (Word64)
+import           Network.ABCI.Types.Messages.FieldTypes
+import           Polysemy                                (Members, Sem,
+                                                          interpret, makeSem)
+import           Polysemy.Error                          (Error)
+import           Tendermint.SDK.BaseApp                  (AppError, ReadStore,
+                                                          WriteStore)
+import qualified Tendermint.SDK.BaseApp.Store.List       as L
+import qualified Tendermint.SDK.BaseApp.Store.Map        as M
+import qualified Tendermint.SDK.BaseApp.Store.Var        as V
+import           Tendermint.SDK.Modules.Validators.Store
+import           Tendermint.SDK.Modules.Validators.Types
+
+
+data ValidatorsKeeper m a where
+  GetValidatorsKeys :: ValidatorsKeeper m (Set.Set PubKey_)
+  GetPowerOf :: PubKey_ -> ValidatorsKeeper m Word64
+  GetQueuedUpdates :: ValidatorsKeeper m (Map.Map PubKey_ Word64)
+  QueueUpdate :: PubKey_ -> Word64 -> ValidatorsKeeper m ()
+
+makeSem ''ValidatorsKeeper
+
+type ValidatorsEffs = '[ValidatorsKeeper]
+
+eval
+  :: Members [ReadStore, WriteStore, Error AppError] r
+  => Sem (ValidatorsKeeper : r) a
+  -> Sem r a
+eval = interpret (\case
+  GetValidatorsKeys -> getValidatorsKeysF
+  GetPowerOf key -> getPowerOfF key
+  GetQueuedUpdates -> getQueuedUpdatesF
+  QueueUpdate key power -> queueUpdateF key power
+  )
+
+getValidatorsKeysF
+  :: Members [ReadStore, Error AppError] r
+  => Sem r (Set.Set PubKey_)
+getValidatorsKeysF =
+  fmap (maybe Set.empty (\(KeySet x) -> x)) $ V.takeVar validatorsKeySet
+
+getPowerOfF
+  :: Members [ReadStore, Error AppError] r
+  => PubKey_
+  -> Sem r Word64
+getPowerOfF key =
+  fmap (fromMaybe 0) $ M.lookup key validatorsMap
+
+getQueuedUpdatesF
+  :: Members [ReadStore, Error AppError] r
+  => Sem r (Map.Map PubKey_ Word64)
+getQueuedUpdatesF = L.foldl (\m (ValidatorUpdate_ ValidatorUpdate{..}) ->
+  Map.alter (Just . fromMaybe (toWord validatorUpdatePower)) (PubKey_ validatorUpdatePubKey) m) Map.empty updatesList
```

Comment:
What does it mean if two ValidatorUpdate_s have been queued for the same PubKey_? It seems like here we just take the one that happened last as the update and disregard whatever came before it?

Reply:
Yes that was the intended behavior here. I figured there were roughly 3 options of how to handle multiple updates to the same validator within a single block:
1. Don't allow it, which would require turning the updates array into a map to check if the pubkey already existed
2. Allow it, and apply only the first update
3. Allow it, and apply only the last update


Applying the last update made the most sense to me, as all the updates are replacements not relative changes. So the result of composing those replacements in order would be to just take the last one.